lib/scenario_runner.py

```diff
@@ -613,6 +614,56 @@ def _populate_image_names(self):
                 else:
                     service['image'] = f"{service_name}_{random.randint(500000,10000000)}"
 
+    def _populate_cpu_and_memory_limits(self):
+        services = self._usage_scenario.get('services', {})
+
+        DOCKER_AVAILABLE_MEMORY = int(subprocess.check_output(['docker', 'info', '--format', '{{.MemTotal}}'], encoding='UTF-8', errors='replace').strip())
+        unassigned_memory = DOCKER_AVAILABLE_MEMORY-1024**3 # we want to leave 1 GB free on the host / docker VM to avoid OOM situations
+
+        if unassigned_memory <= 0:
+            raise RuntimeError(f"Docker has insufficient memory available. Available: {DOCKER_AVAILABLE_MEMORY/1024**3:.2f}GB, Required: at least 1GB for GMT overhead")
+
+        SYSTEM_ASSIGNABLE_CPU_COUNT = int(subprocess.check_output(['docker', 'info', '--format', '{{.NCPU}}'], encoding='UTF-8', errors='replace').strip()) -1
+        if SYSTEM_ASSIGNABLE_CPU_COUNT <= 0:
+            raise RuntimeError(f"Cannot assign docker containers to any CPU as no CPUs are available to Docker. Available CPU count: {SYSTEM_ASSIGNABLE_CPU_COUNT}")
+
+        to_be_assigned_services = []
+        for service_name, service in services.items():
+            # wildly the docker compose spec allows deploy to be None
+            # ... thus we need to check and cannot .get()
+            if 'deploy' in service and service['deploy'] is not None and (memory := service['deploy'].get('resources', {}).get('limits', {}).get('memory', None)) is not None:
+                del service['deploy']['resources']['limits']['memory']
+                service['mem_limit'] = memory
+            if service.get('mem_limit', 0) == 0:
+                to_be_assigned_services.append(service_name)
+                self.__warnings.append(f"Service '{service_name}' had no memory limit set. GMT does not allow unbounded memory limits and auto value was applied.")
```

Comment:
This warning message will be displayed for a lot of runs. Maybe it would make sense to add a link to a documentation page, e.g. https://docs.docker.com/reference/compose-file/deploy/#resources